File: src/geeks_for_geeks/ds/graph/edge_repr/Graph.java
package geeks_for_geeks.ds.graph.edge_repr;

import geeks_for_geeks.ds.nodes.Edge;
import geeks_for_geeks.ds.union_find.UnionFind;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * This is undirected graph.
 * Created By: Prashant Chaubey
 * Created On: 15-10-2018 22:40
 **/
public class Graph {
    private List<Edge> edges;
    private ArrayList<Integer>[] values;
    private int vertices;

    public Graph(int vertices) {
        this.vertices = vertices;
        this.edges = new ArrayList<>();
        this.values = new ArrayList[vertices];
        for (int i = 0; i < vertices; i++) {
            values[i] = new ArrayList<>();
        }
    }


    public Graph addEdge(int src, int dest, int weight) {
        assert dest >= 0 && dest < vertices && src >= 0 && src < vertices;
        int index = edges.size();
        this.values[src].add(index);
        this.values[dest].add(index);
        this.edges.add(new Edge(src, dest, weight));
        return this;
    }

    public boolean isCyclic() {
        UnionFind uf = new UnionFind(this.vertices);

        for (Edge edge : edges) {
            int parent1 = uf.find(edge.src);
            int parent2 = uf.find(edge.dest);

            if (parent1 == parent2) {
                return true;
            }

            uf.union(edge.src, edge.dest);
        }

        return false;
    }


    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Graph{");
        for (Edge edge : edges) {
            sb.append(edge).append(",").append(System.lineSeparator());
        }
        sb.append("}");
        return sb.toString();
    }

    public boolean isConnected(int source) {
        boolean[] visited = new boolean[this.vertices];
        Arrays.fill(visited, false);
        dfsUtil(source, visited);
        for (boolean i : visited) {
            if (!i) {
                return false;
            }
        }
        return true;
    }

    public boolean isEulerianCircuitExists() {
        if (!isConnected(0)) {
            return false;
        }
        for (int i = 0; i < this.values.length; i++) {
            if (this.values[i].size() % 2 == 1) {
                return false;
            }
        }
        return true;
    }

    public void dfsUtil(int source, boolean[] visited) {
        visited[source] = true;
        ArrayList<Integer> neighbours = this.values[source];
        for (Integer neighbourEdge : neighbours) {
            Edge edge = this.edges.get(neighbourEdge);
            int neighbour = source ^ edge.src ^ edge.dest;
            if (!visited[neighbour]) {
                dfsUtil(neighbour, visited);
            }
        }
    }

}

File: src/main/java/gfg/ds/graph/graph.py
class Graph:
    def __init__(self, vertices):
        self.values = [-1] * vertices
        for i in range(vertices):
            self.values[i] = []
        self.in_degree = [0] * vertices
        self.vertex_exists = [False] * vertices

    def add_edge(self, src, dest):
        self.values[src].append(dest)
        self.in_degree[dest] += 1
        self.vertex_exists[dest] = self.vertex_exists[src] = True

    def dfs_util(self, src, visited):
        visited[src] = True
        neighbours = self.values[src]
        for i in neighbours:
            if not visited[i]:
                self.dfs_util(i, visited)

    def transpose(self):
        vertices = len(self.values)
        transpose = Graph(vertices)
        for i in range(vertices):
            neighbours = self.values[i]
            for neighbour in neighbours:
                transpose.add_edge(neighbour, i)
        return transpose

    def is_all_vertices_reachable(self, source):
        vertices = len(self.values)
        visited = [False] * vertices
        self.dfs_util(source, visited)
        for i in range(vertices):
            if self.vertex_exists[i] and not visited[i]:
                return False
        return True

    def is_eulerian_path_exists(self):
        is_out_greater_found = False
        source = -1
        is_in_greater_found = False
        vertices = len(self.values)
        for i in range(vertices):
            out_degree = len(self.values[i])
            in_degree = self.in_degree[i]
            if out_degree > in_degree:
                if out_degree - in_degree == 1 and not is_out_greater_found:
                    is_out_greater_found = True
                    source = i
                else:
                    return False
            elif in_degree > out_degree:
                if in_degree - out_degree == 1 and not is_in_greater_found:
                    is_in_greater_found = True
                else:
                    return False
        if is_in_greater_found and is_out_greater_found:
            return self.is_all_vertices_reachable(source)
        else:
            for i in range(vertices):
                if self.vertex_exists[i] and len(self.values[i]) > 0:
                    return self.is_all_vertices_reachable(i)
        assert False
